contracts/protocore/Protocore.sol

```diff
@@ -310,6 +334,226 @@ contract Protocore is MosaicVersion, CoconsensusModule {
+        require(
+            _voteMessageHash != bytes32(0),
+            "Vote message hash is 0."
+        );
+
+        Link storage link = links[_voteMessageHash];
+
+        require(
+            link.targetFinalisation >= CheckpointFinalisationStatus.Registered,
+            "The given link status is at least reported."
+        );
+
+        require(
+            link.proposedMetablockHeight.add(1) >= openKernelHeight,
+            "Link height inclusion principle has surpassed."
+        );
+
+        address validator = ecrecover(_voteMessageHash, _v, _r, _s);
+
+        require(
+            validator != address(0),
+            "Validator must not be null."
+        );
+
+        require(
+            fvsVotes[_voteMessageHash][openKernelHeight][validator] == address(0),
+            "Validator vote cannot be registered twice in FVS votes at the same height."
+        );
+
+        bool quorumReached = countVoteForForwardValidatorSets(_voteMessageHash, link, validator);
+        insertForwardValidatorVote(_voteMessageHash, validator);
+
+        emit VoteRegistered(_voteMessageHash, openKernelHeight, _r, _s, _v);
+
+        if (quorumReached) {
+            justifyLink(_voteMessageHash, link);
+        }
+
+    }
+
+
+    /* Private Functions */
+
+    /**
+     * @notice calculateQuorum() function calculates quorum of the given
+     *         count.
+     *         The super majority numerator and dominator must be the same
+     *         as in the corresponding core.
+     */
+    function calculateQuorum(uint256 _count)
+        private
+        pure
+        returns (uint256 quorum_)
+    {
+        quorum_ = _count
+            .mul(CORE_SUPER_MAJORITY_NUMERATOR)
+            .div(CORE_SUPER_MAJORITY_DENOMINATOR);
+    }
+
+    function countVoteForForwardValidatorSets(
+        bytes32 _voteMessageHash,
+        Link storage _link,
+        address _validator
+    )
+        private
+        returns (bool quorumReached_)
+    {
+        bool quorumForwardValidatorSet = false;
+        if (canVoteInForwardValidatorSet(
+            _voteMessageHash,
+            openKernelHeight,
+            _validator)) {
+            _link.fvsVoteCount[openKernelHeight] = _link.fvsVoteCount[openKernelHeight].add(1);
+            quorumForwardValidatorSet = (_link.fvsVoteCount[openKernelHeight] >= fvsQuorums[openKernelHeight]);
+        }
+
+        bool quorumRearValidatorSet = true;
+        if (openKernelHeight > uint256(0)) {
+            uint256 previousHeight = openKernelHeight.sub(1);
+            if (canVoteInForwardValidatorSet(
+                _voteMessageHash,
+                previousHeight,
+                _validator)) {
+                _link.fvsVoteCount[openKernelHeight] = _link.fvsVoteCount[previousHeight].add(1);
+            }
+            quorumRearValidatorSet = (_link.fvsVoteCount[previousHeight] >= fvsQuorums[previousHeight]);
+        }
+        quorumReached_ = (quorumForwardValidatorSet && quorumRearValidatorSet);
```

Comment:
`quorumForwardValidatorSet` is `false` by default.
If the validator cannot vote in forward validator set but can vote in rear validator set then  `quorumForwardValidatorSet` will remain false.
`quorumReached_ ` will always be `false`.

Should the following code at line 464 be taken outside the if condition?
```
quorumForwardValidatorSet = (_link.fvsVoteCount[openKernelHeight] >= fvsQuorums[openKernelHeight]);
```

Reply:
Code deduplicate: The code logic is repeated in both the if condition, this can be extracted in a function

Reply:
good catch; yes, L464 should be outside of the `canVoteInValidatorSet`.

I would not abstract `_link.fvsVoteCount[openKernelHeight] >= fvsQuorums[openKernelHeight]` because now it's inline.

Reply:
Done.

Reply:
LGTM